### PR TITLE
Add function ULIDStringToDateTime

### LINKED ISSUE #46945: Function for decoding ULIDs
Support for generating a ULID was added very recently here: https://github.com/ClickHouse/ClickHouse/pull/44662. It would be great if we can also decode them.

**Use case**

Derive the timestamp that is contained within a ULID, so that it be used in queries, used in a (materialized) column etc.

**Describe the solution you'd like**

Add function decodeULID that returns a DateTime.

**Describe alternatives you've considered**

-

**Additional context**

https://github.com/ClickHouse/ulid-c already has a function that decodes a ULID.

### PR DESCRIPTION
**Pull request by evillique**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add function ULIDStringToDateTime(). Closes https://github.com/ClickHouse/ClickHouse/issues/46945

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

### CODE REVIEW

File: docs/en/sql-reference/functions/ulid-functions.md [hunk 1]
```diff
@@ -48,6 +48,39 @@ SELECT generateULID(1), generateULID(2)
...
+
+**Syntax**
+
+``` sql
+ULIDStringToDateTime(ulid[, timezone])
+```
+
+**Arguments**
+
+-   `ulid` — Input UUID. [String](/docs/en/sql-reference/data-types/string.md) or [FixedString(26)](/docs/en/sql-reference/data-types/fixedstring.md).
```
Comment (Avogar) [inline docs/en/sql-reference/functions/ulid-functions.md:63]:
```suggestion
-   `ulid` — Input ULID. [String](/docs/en/sql-reference/data-types/string.md) or [FixedString(26)](/docs/en/sql-reference/data-types/fixedstring.md).
```

File: src/Functions/FunctionsCodingULID.cpp [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+            size_t src_offset = 0;
+
+            for (size_t i = 0; i < input_rows_count; ++i)
+            {
+                DateTime64 time = 0;
+
+                size_t string_size = offsets_src[i] - src_offset;
+                if (string_size == ULID_LENGTH + 1)
+                    time = decode(vec_src.data() + src_offset);
```
Comment (Avogar) [inline src/Functions/FunctionsCodingULID.cpp:124]: So, if string size is not `ULID_LENGTH + 1` we just return default datetime? I think it's unexpected and It's better to throw an exception in this case. If we want to get default value or Null on bad ULID, we can implement functions `ULIDStringToDateTimeOrNull/ULIDStringToDateTimeOrDefault`

File: tests/queries/0_stateless/02668_ulid_decoding.sql [hunk 3]
```diff
@@ -0,0 +1,3 @@
+-- Tags: no-fasttest
+
+SELECT dateDiff('s', ULIDStringToDateTime(generateULID()), now()) = 0;
```
Comment (Avogar) [inline tests/queries/0_stateless/02668_ulid_decoding.sql:3]: Let's also add tests for invalid arguments, wrong arguments types and invalid ULID (with different sizes)

**Review by Avogar — APPROVED** [review]

**Comment by evillique** [comment]
Test failures:
- Integration tests (asan) [1/6]: https://github.com/ClickHouse/ClickHouse/issues/48180
- Stress test (asan): https://github.com/ClickHouse/ClickHouse/issues/45372

**Pull request merged**